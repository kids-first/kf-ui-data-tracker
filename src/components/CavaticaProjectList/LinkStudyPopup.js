import React, {useState} from 'react';
import {Button, Popup} from 'semantic-ui-react';
import {useQuery, useMutation} from '@apollo/react-hooks';
import {Formik} from 'formik';
import {ALL_STUDIES, GET_PROJECTS} from '../../state/queries';
import {LINK_PROJECT} from '../../state/mutations';
import {LinkStudyForm} from '../../forms';

export const LinkStudyPopup = ({project}) => {
  const [isOpen, setOpen] = useState(false);

  const {loading, data, error} = useQuery(ALL_STUDIES, {
    fetchPolicy: 'cache-first',
  });

  const [linkProject, {loading: linkLoading}] = useMutation(LINK_PROJECT, {
    refetchQueries: [
      {
        query: GET_PROJECTS,
      },
    ],
  });

  if (loading || !data || error) {
    return <></>;
  }

  const allStudies = data && data.allStudies;

  return (
    <Popup
      on="click"
      position="left center"
      wide="very"
      open={isOpen}
      onClose={() => setOpen(false)}
      onOpen={() => setOpen(true)}
      trigger={
        <Button primary basic size="mini" icon="chain" content="LINK STUDY" />
      }
    >
      <Formik
        initialValues={{
          studyId: null,
        }}
        validate={values => {
          let errors = {};
          if (!values.studyId) {
            errors.studyId = 'Required';
          }
          return errors;
        }}
        onSubmit={values => {
          linkProject({
            variables: {
              project: project.id,
              study: values.studyId,
            },
          }).then(resp => setOpen(false));
        }}
      >
        {formikProps => (
          <LinkStudyForm
            allStudies={allStudies}
            formikProps={formikProps}
            loading={linkLoading}
          />
        )}
      </Formik>
    </Popup>
  );
};

export default LinkStudyPopup;
