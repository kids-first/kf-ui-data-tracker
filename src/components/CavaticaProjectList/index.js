export {default as CavaticaProjectList} from './CavaticaProjectList';
