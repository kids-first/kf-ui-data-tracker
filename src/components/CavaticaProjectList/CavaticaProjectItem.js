import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {
  List,
  Icon,
  Button,
  Confirm,
  Message,
  Modal,
  Popup,
  Divider,
  Header,
} from 'semantic-ui-react';
import TimeAgo from 'react-timeago';
import {longDate} from '../../common/dateUtils';
import {projectOptions} from '../../common/enums';
import {EditProjectModal} from '../../modals';
import {LinkStudyPopup} from './LinkStudyPopup';

const ProjectAttributes = ({projectNode, disabled}) => (
  <List bulleted horizontal>
    <List.Item
      as={Header}
      sub
      disabled={disabled}
      color={
        projectOptions.find(i => i.key === projectNode.projectType).color ||
        'grey'
      }
    >
      {projectOptions.find(i => i.key === projectNode.projectType).text ||
        'Unknown'}
    </List.Item>
    <List.Item disabled={disabled}>
      Created
      {projectNode.createdBy ? ' by ' + projectNode.createdBy + ' ' : ' '}
      <TimeAgo
        live={false}
        date={projectNode.createdOn}
        title={longDate(projectNode.createdOn)}
      />
    </List.Item>
    {projectNode.workflowType && (
      <List.Item disabled={disabled}>{projectNode.workflowType}</List.Item>
    )}
    <List.Item disabled={disabled}>
      <code>{projectNode.projectId}</code>
    </List.Item>
  </List>
);

const ProjectLink = ({projectNode, disableLink}) => {
  if (disableLink) {
    return <List.Header>{projectNode.name}</List.Header>;
  } else {
    return (
      <List.Header
        as="a"
        target="_blank"
        href={`https://cavatica.sbgenomics.com/u/${projectNode.projectId}`}
      >
        {projectNode.name + ' '}
        <Icon link size="small" name="external" />
      </List.Header>
    );
  }
};

const UnlinkButton = ({unlinkProject, study, projectId}) => (
  <Popup
    trigger={
      <Button
        basic
        negative
        floated="right"
        size="mini"
        icon="unlink"
        content="UNLINK"
        className="ml-10"
        onClick={e => e.stopPropagation()}
      />
    }
    header="Are you sure?"
    content={
      <>
        This will unlink the project from its study. It may always be linked
        back later.
        <Divider />
        <Button
          data-testid="delete-confirm"
          negative
          fluid
          icon={<Icon name="unlink" />}
          content="Unlink"
          onClick={e => {
            e.stopPropagation();
            unlinkProject({
              variables: {
                project: projectId,
                study: study.id,
              },
            });
          }}
        />
      </>
    }
    on="click"
    position="top right"
  />
);

export const ImportVolumeButton = ({importVolumeFiles, projectNode}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  // Starts from 'confirm' then progresses to 'success'
  const [state, setState] = useState('confirm');

  const handleSubmit = e => {
    if (state === 'success') {
      setIsOpen(false);
      setState('confirm');
      setError('');
      setLoading(false);
      return;
    }
    setLoading(true);

    e.stopPropagation();
    importVolumeFiles({
      variables: {
        project: projectNode.id,
      },
    })
      .then(resp => {
        if (resp.errors) {
          setError(JSON.stringify(resp.errors));
        }
        setLoading(false);
        setState('success');
      })
      .catch(err => {
        setError(err);
        setLoading(false);
      });
  };

  const directory = new Date().toISOString().slice(0, 10) + '-kf-data-delivery';

  return (
    <>
      <Button
        primary
        className="ml-10"
        icon={<Icon name="cloud download" />}
        content="Import Volume"
        labelPosition="left"
        size="mini"
        onClick={() => setIsOpen(true)}
      />

      <Confirm
        open={isOpen}
        header="Import volume files to Cavatica project"
        onCancel={() => {
          setState('confirm');
          setError('');
          setIsOpen(false);
        }}
        onConfirm={handleSubmit}
        confirmButton={
          loading ? (
            <Button loading disabled content="submitting" />
          ) : state === 'confirm' ? (
            'Proceed with Import'
          ) : (
            'OK'
          )
        }
        cancelButton={state === 'confirm' ? 'Cancel' : null}
        content={
          state === 'confirm' ? (
            <Modal.Content>
              <p>
                You are about to import all files from the connected study
                volume's <b>source/</b> prefix into this delivery project. Any
                users that have access to this project will be able to view
                these files.
              </p>
              <p>Here's what's about to happen:</p>
              <List bulleted>
                <List.Item>
                  A new folder named <b>{directory}</b> will be created in the
                  project
                </List.Item>
                <List.Item>
                  All files inside the <b>source/</b> prefix of the study volume
                  will be imported to this folder.
                </List.Item>
                <List.Item>
                  All files uploaded through the Data Tracker under{' '}
                  <b>source/uploads/</b> will <b>not</b> be imported
                </List.Item>
                <List.Item>
                  If the volume has already been imported to this project today,
                  the previous import will be <b>overwritten</b>
                </List.Item>
              </List>
              {error && <Message error header="Error" content={error} />}
            </Modal.Content>
          ) : (
            <Modal.Content>
              <Message
                success
                icon={<Icon size="mini" name="warning" />}
                header="Import Started"
                content={
                  <>
                    <p>Files may take a couple minutes to copy completely.</p>
                    <p>
                      Please view the{' '}
                      <a
                        as="a"
                        target="_blank"
                        rel="noopener noreferrer"
                        href={`https://cavatica.sbgenomics.com/u/${
                          projectNode.projectId
                        }/files/#q?path=${directory}`}
                      >
                        project's file directory{' '}
                        <Icon link size="small" name="external" />
                      </a>{' '}
                      and review it for any files that may not belong.
                    </p>
                  </>
                }
              />
            </Modal.Content>
          )
        }
      />
    </>
  );
};

const CavaticaProjectItem = ({
  projectNode,
  unlinkProject,
  importVolumeFiles,
  disableLink,
  editable = true,
  showLinkButton,
}) => {
  const [isEditing, setIsEditing] = useState(false);

  if (projectNode.deleted) {
    return (
      <List.Item>
        <List.Content floated="right" verticalAlign="middle">
          {unlinkProject && projectNode.study && (
            <UnlinkButton
              unlinkProject={unlinkProject}
              projectId={projectNode.id}
              study={projectNode.study}
            />
          )}
          {projectNode.projectType === 'DEL' && importVolumeFiles && (
            <ImportVolumeButton
              importVolumeFiles={importVolumeFiles}
              projectNode={projectNode}
            />
          )}
        </List.Content>
        <List.Content>
          <Header floated="right" disabled size="tiny">
            Deleted
          </Header>
          <List.Header as={Header} color="grey" size="tiny">
            {projectNode.name + ' '}
          </List.Header>
          <ProjectAttributes projectNode={projectNode} disabled />
        </List.Content>
      </List.Item>
    );
  } else {
    return (
      <>
        <List.Item>
          <List.Content floated="right">
            {projectNode.projectType === 'DEL' && importVolumeFiles && (
              <ImportVolumeButton
                importVolumeFiles={importVolumeFiles}
                projectNode={projectNode}
              />
            )}
            {editable && (
              <Popup
                trigger={
                  <Button
                    primary
                    basic
                    size="mini"
                    floated="right"
                    icon="pencil"
                    content="EDIT"
                    onClick={() => setIsEditing(true)}
                  />
                }
                inverted
                content="Edit this project"
                on="hover"
                position="top right"
              />
            )}
            {!projectNode.study && showLinkButton && (
              <LinkStudyPopup project={projectNode} />
            )}
            {unlinkProject && projectNode.study && (
              <UnlinkButton
                unlinkProject={unlinkProject}
                projectId={projectNode.id}
                study={projectNode.study}
              />
            )}
          </List.Content>
          <List.Content>
            <ProjectLink projectNode={projectNode} disableLink={disableLink} />
            <ProjectAttributes projectNode={projectNode} />
          </List.Content>
        </List.Item>
        <EditProjectModal
          open={editable && isEditing}
          projectNode={projectNode}
          onCloseDialog={() => setIsEditing(false)}
        />
      </>
    );
  }
};

CavaticaProjectItem.propTypes = {
  /** Project object with study object nested if linked */
  projectNode: PropTypes.object.isRequired,
  /** Action to unlink a project taking project id and study id */
  unlinkProject: PropTypes.func,
  /** If disable the external link to Cavatica on project name */
  disableLink: PropTypes.bool,
  /** Whether or not to allow the edit modal to be spawned */
  editable: PropTypes.bool,
};

export default CavaticaProjectItem;
