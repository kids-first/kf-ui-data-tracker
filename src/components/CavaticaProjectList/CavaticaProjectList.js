import React from 'react';
import PropTypes from 'prop-types';
import {List} from 'semantic-ui-react';
import CavaticaProjectItem from './CavaticaProjectItem';
/**
 * Displays Cavatica projects in a list format with the option to
 * show/hide edit and link button, and study links
 */
const CavaticaProjectList = ({
  projects,
  unlinkProject,
  importVolumeFiles,
  editable,
  disableLink,
  showLinkButton,
}) => (
  <List relaxed divided>
    {projects &&
      projects.length > 0 &&
      projects
        .sort(({node: p1}, {node: p2}) => {
          if (p1.deleted !== p2.deleted) {
            // Sort deleted projects to the bottom
            return p1.deleted ? 1 : -1;
          } else if (!p1.study || !p2.study) {
            if (p1.study === p2.study) {
              // Sort by project name
              return p1.projectId.localeCompare(p2.projectId);
            }
            // Sort unlinked projects toward the bottom
            if (p1.study === null) {
              return 1;
            } else if (p2.study === null) {
              return -1;
            }
          }
          // Sort by the linked study's name
          return p1.study.name.localeCompare(p2.study.name);
        })
        .map(({node}) => (
          <CavaticaProjectItem
            key={node.id}
            projectNode={node}
            unlinkProject={unlinkProject}
            editable={editable}
            disableLink={disableLink}
            importVolumeFiles={importVolumeFiles}
            showLinkButton={showLinkButton}
          />
        ))}
  </List>
);

CavaticaProjectList.propTypes = {
  /** Array of Cavatica projects object*/
  projects: PropTypes.array,
  /** Function to unlink the project from the study */
  unlinkProject: PropTypes.func,
  /** If allow user to see the edit button */
  editable: PropTypes.bool,
  /** If disable the link going to Cavatica */
  disableLink: PropTypes.bool,
};

export default CavaticaProjectList;
