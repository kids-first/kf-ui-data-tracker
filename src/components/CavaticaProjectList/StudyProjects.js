import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import {Accordion, Divider, Header, Icon, List} from 'semantic-ui-react';
import CavaticaProjectItem from './CavaticaProjectItem';

const projectSort = ({node: p1}, {node: p2}) => {
  if (p1.deleted !== p2.deleted) {
    // Sort deleted projects to the bottom
    return p1.deleted ? 1 : -1;
  } else if (!p1.study || !p2.study) {
    if (p1.study === p2.study) {
      // Sort by project name
      return p1.projectId.localeCompare(p2.projectId);
    }
    // Sort unlinked projects toward the bottom
    if (p1.study === null) {
      return 1;
    } else if (p2.study === null) {
      return -1;
    }
  }
  // Sort by the linked study's name
  return p1.study.name.localeCompare(p2.study.name);
};

/**
 * Display a list of Cavatica projects for a given study
 */
const StudyProjects = ({
  study,
  projects,
  unlinkProject,
  importVolumeFiles,
  editable,
  disableLink,
}) => {
  const [active, setActive] = useState(!study);

  const handleClick = (e, title) => {
    console.log(active);
    setActive(!active);
  };

  return (
    <>
      <Accordion.Title active={active} onClick={handleClick}>
        <Header size="medium">
          <Icon name="dropdown" />
          {study ? (
            <>
              ({projects.length}){' '}
              <Link to={`/study/${study.kfId}/cavatica`}>{study.kfId}</Link> -{' '}
              {study.name}
            </>
          ) : (
            <>({projects.length}) Unlinked Projects</>
          )}
          <Header sub floated="right">
            Click to {active ? 'Collapse' : 'Expand'}
          </Header>
        </Header>
      </Accordion.Title>
      <Accordion.Content active={active}>
        <List relaxed divided>
          {projects &&
            projects.length > 0 &&
            projects
              .sort(projectSort)
              .map(({node}) => (
                <CavaticaProjectItem
                  key={node.id}
                  projectNode={node}
                  unlinkProject={unlinkProject}
                  editable={editable}
                  disableLink={disableLink}
                  importVolumeFiles={importVolumeFiles}
                />
              ))}
        </List>
      </Accordion.Content>
      <Divider />
    </>
  );
};

StudyProjects.propTypes = {
  /** The study node */
  study: PropTypes.object,
  /** Array of Cavatica projects object*/
  projects: PropTypes.array,
  /** Function to unlink the project from the study */
  unlinkProject: PropTypes.func,
  /** If allow user to see the edit button */
  editable: PropTypes.bool,
  /** If disable the link going to Cavatica */
  disableLink: PropTypes.bool,
};

export default StudyProjects;
