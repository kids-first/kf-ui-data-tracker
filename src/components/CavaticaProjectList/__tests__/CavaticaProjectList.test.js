import React from 'react';
import {render} from 'react-testing-library';
import {MemoryRouter} from 'react-router-dom';
import CavaticaProjectList from '../CavaticaProjectList';

it('renders correctly', () => {
  const projects = [
    {
      node: {
        id:
          'UHJvamVjdE5vZGU6a29sYm1hbmQvc2Qtd2dwOHJ3M3ctZ2F0ay1oYXBsb3R5cGVjYWxsZXI=',
        createdBy: 'kolbmand',
        createdOn: '2019-08-14T18:29:01+00:00',
        projectId: 'kolbmand/sd-wgp8rw3w-gatk-haplotypecaller',
        projectType: 'DEL',
        name: 'SD_WGP8RW3W gatk-haplotypecaller',
        workflowType: 'bwa_mem',
        study: {
          id: 'U3R1ZHlOb2RlOlNEX1dHUDhSVzNX',
          kfId: 'SD_WGP8RW3W',
          name: null,
          shortName: null,
        },
      },
    },
    {
      node: {
        id: 'UHJvamVjdE5vZGU6a2ZkcmMtaGFybW9uaXphdGlvbi9zZC1lMWJyMnRmaw==',
        createdBy: 'kfdrc-harmonization',
        createdOn: '2019-06-10T14:49:01+00:00',
        projectId: 'kfdrc-harmonization/sd-e1br2tfk',
        projectType: 'DEL',
        name: 'data harmonization',
        workflowType: 'bwa_mem',
        study: null,
      },
    },
    {
      node: {
        id: 'UHJvamVjdE5vZGU6a2ZkcmMtaGFybW9uaXphdGlvbi9zZC1qd3MzdjI0ZC0wMQ==',
        createdBy: 'kfdrc-harmonization',
        createdOn: '2019-05-21T14:13:10+00:00',
        projectId: 'kfdrc-harmonization/sd-jws3v24d-01',
        projectType: 'DEL',
        name: 'Test Joint Genotyping',
        workflowType: 'bwa_mem',
        study: null,
      },
    },
  ];
  const tree = render(
    <MemoryRouter initialEntries={['/cavatica-projects']}>
      <CavaticaProjectList projects={projects} />
    </MemoryRouter>,
  );
  expect(tree.container).toMatchSnapshot();
});
