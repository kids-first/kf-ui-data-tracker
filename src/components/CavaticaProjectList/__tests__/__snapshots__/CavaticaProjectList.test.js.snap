// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cavatica project list correctly - default look 1`] = `
<div>
  <div
    class="ui divided relaxed list"
    role="list"
  >
    <div
      class="item"
      role="listitem"
    >
      <div
        class="right floated content"
      >
        <a
          href="/study/SD_WGP8RW3W/basic-info/info"
        >
          Test Study
        </a>
        <button
          class="ui mini basic primary right floated button"
        >
          <i
            aria-hidden="true"
            class="pencil icon"
          />
          EDIT
        </button>
      </div>
      <i
        aria-hidden="true"
        class="paper plane outline icon"
      />
      <div
        class="content"
      >
        <a
          class="header"
          href="https://cavatica.sbgenomics.com/u/kfdrc-harmonization/sd-e1br2tfk"
          target="_blank"
        >
          data harmonization 
          <i
            aria-hidden="true"
            class="external small link icon"
          />
        </a>
        <div
          class="ui bulleted horizontal list"
          role="list"
        >
          <div
            class="item"
            role="listitem"
          >
            Created
             by kfdrc-harmonization 
            <time
              datetime="2019-06-10T14:49:01.000Z"
              title="10 Jun 2019"
            >
              2 months from now
            </time>
          </div>
          <div
            class="item"
            role="listitem"
          >
            bwa_mem
          </div>
          <div
            class="item"
            role="listitem"
          >
            <code>
              kfdrc-harmonization/sd-e1br2tfk
            </code>
          </div>
        </div>
      </div>
    </div>
    <div
      class="item"
      role="listitem"
    >
      <div
        class="right floated content"
      >
        Not linked
        <button
          class="ui mini basic primary right floated button"
        >
          <i
            aria-hidden="true"
            class="pencil icon"
          />
          EDIT
        </button>
      </div>
      <i
        aria-hidden="true"
        class="sliders horizontal icon"
      />
      <div
        class="content"
      >
        <a
          class="header"
          href="https://cavatica.sbgenomics.com/u/kfdrc-harmonization/sd-jws3v24d-01"
          target="_blank"
        >
          Test Joint Genotyping 
          <i
            aria-hidden="true"
            class="external small link icon"
          />
        </a>
        <div
          class="ui bulleted horizontal list"
          role="list"
        >
          <div
            class="item"
            role="listitem"
          >
            Created
             by kfdrc-harmonization 
            <time
              datetime="2019-05-21T14:13:10.000Z"
              title="21 May 2019"
            >
              4 weeks from now
            </time>
          </div>
          <div
            class="item"
            role="listitem"
          >
            bwa_mem
          </div>
          <div
            class="item"
            role="listitem"
          >
            <code>
              kfdrc-harmonization/sd-jws3v24d-01
            </code>
          </div>
        </div>
      </div>
    </div>
    <div
      class="item"
      role="listitem"
    >
      <div
        class="right floated content"
      >
        Not linked
        <button
          class="ui mini basic primary right floated button"
        >
          <i
            aria-hidden="true"
            class="pencil icon"
          />
          EDIT
        </button>
      </div>
      <i
        aria-hidden="true"
        class="sliders horizontal icon"
      />
      <div
        class="content"
      >
        <a
          class="header"
          href="https://cavatica.sbgenomics.com/u/kfdrc-harmonization/sd-jws3v24d-bwa-mem"
          target="_blank"
        >
          Testing 
          <i
            aria-hidden="true"
            class="external small link icon"
          />
        </a>
        <div
          class="ui bulleted horizontal list"
          role="list"
        >
          <div
            class="item"
            role="listitem"
          >
            Created
             by kfdrc-harmonization 
            <time
              datetime="2019-05-21T14:13:00.000Z"
              title="21 May 2019"
            >
              4 weeks from now
            </time>
          </div>
          <div
            class="item"
            role="listitem"
          >
            bwa_mem
          </div>
          <div
            class="item"
            role="listitem"
          >
            <code>
              kfdrc-harmonization/sd-jws3v24d-bwa-mem
            </code>
          </div>
        </div>
      </div>
    </div>
    <div
      class="item"
      role="listitem"
    >
      <div
        class="right floated middle aligned content"
      />
      <i
        aria-hidden="true"
        class="grey paper plane outline icon"
      />
      <div
        class="content"
      >
        <div
          class="ui tiny disabled right floated header"
        >
          Deleted
        </div>
        <div
          class="ui grey tiny header header"
        >
          SD_WGP8RW3W gatk-haplotypecaller 
        </div>
        <div
          class="ui bulleted horizontal list"
          role="list"
        >
          <div
            class="disabled item"
            role="listitem"
          >
            Created
             by kolbmand 
            <time
              datetime="2019-08-14T18:29:01.000Z"
              title="14 Aug 2019"
            >
              4 months from now
            </time>
          </div>
          <div
            class="disabled item"
            role="listitem"
          >
            bwa_mem
          </div>
          <div
            class="disabled item"
            role="listitem"
          >
            <code>
              kolbmand/sd-wgp8rw3w-gatk-haplotypecaller
            </code>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders cavatica project list correctly - no data 1`] = `
<div>
  <div
    class="ui divided relaxed list"
    role="list"
  />
</div>
`;
